clang/lib/CConv/CastPlacement.cpp

```diff
@@ -110,49 +87,49 @@ bool CastPlacementVisitor::needCasting(ConstraintVariable *Src,
   auto &E = Info.getConstraints().getVariables();
-  auto SrcChecked = Src->isChecked(E);
   // Check if the src is a checked type.
-  if (SrcChecked) {
-    // Check if Dst is an itype, if yes then
-    // Src should have exactly same checked type else we need to insert cast.
-    if (Dst->hasItype()) {
-    return !Dst->solutionEqualTo(Info.getConstraints(), Src);
-    }
+  if (Src->isChecked(E)) {
+    // If Dst has an itype, Src must have exactly the same checked type. If this
+    // is not the case, we must insert a case.
+    if (Dst->hasItype())
+      return !Dst->solutionEqualTo(Info.getConstraints(), Src);
 
     // Is Dst Wild?
-    if (!Dst->isChecked(E) || Dinfo == WILD) {
+    if (!Dst->isChecked(E) || Dinfo == WILD)
       return true;
-    }
-
-  } return false; }
+  }
+  return false;
+}
 
 // Get the type name to insert for casting.
 std::string CastPlacementVisitor::getCastString(ConstraintVariable *Src,
                                                 ConstraintVariable *Dst,
                                                 IsChecked Dinfo) {
   assert(needCasting(Src, Dst, Dinfo) && "No casting needed.");
-  // The destination type should be a non-checked type.
-  // This is not necessary because of itypes
-  //auto &E = Info.getConstraints().getVariables();
-  //assert(!Dst->anyChanges(E) || Dinfo == WILD);
-  return "((" + Dst->getRewritableOriginalTy() + ")";
+  return "(" + Dst->getRewritableOriginalTy() + ")";
 }
 
-void CastPlacementVisitor::surroundByCast(std::string CastPrefix, Expr *E) {
-  if (Writer.InsertTextAfterToken(E->getEndLoc(), ")")) {
+void CastPlacementVisitor::surroundByCast(const std::string &CastPrefix,
+                                          Expr *E) {
+  // If E is already a cast expression, we will try to rewrite the cast instead
+  // adding a new expression.
+  if (auto *CE = dyn_cast<CStyleCastExpr>(E->IgnoreParens())) {
+    SourceRange CastTypeRange(CE->getLParenLoc(), CE->getRParenLoc());
+    Writer.ReplaceText(CastTypeRange, CastPrefix);
+  } else if (Writer.InsertTextAfterToken(E->getEndLoc(), ")")) {
     // This means we failed to insert the text at the end of the RHS.
     // This can happen because of Macro expansion.
     // We will see if this is a single expression statement?
     // If yes, then we will use parent statement to add ")"
     auto CRA = CharSourceRange::getTokenRange(E->getSourceRange());
     auto NewCRA = clang::Lexer::makeFileCharRange(CRA,
-                                                Context->getSourceManager(),
+                                                  Context->getSourceManager(),
                                                   Context->getLangOpts());
     std::string SrcText = clang::tooling::getText(CRA, *Context);
     // Only insert if there is anything to write.
     if (!SrcText.empty())
-      Writer.ReplaceText(NewCRA, CastPrefix + SrcText + ")");
+      Writer.ReplaceText(NewCRA, "(" + CastPrefix + SrcText + ")");
```

Comment:
I'm surprised about this. I think @Machiry skipped the parenthesis for a reason, but I don't remember what it was. Perhaps you adjusted the generation of `CastPrefix` above so that the parenthesis is now needed, here?

Reply:
This was just some reorganization of existing code. The left parentheses was added somewhere else, but I think it makes more sense to do it here.